Documentation/Evolution/UnicodeForStringProcessing.md

```diff
@@ -0,0 +1,758 @@
+```
+
+Swift's `Regex` follows the level 2 guidelines for Unicode support in regular expressions described in [Unicode Technical Standard #18][uts18], with support for Unicode character classes, canonical equivalence, grapheme cluster matching semantics, and level 2 word boundaries enabled by default. `Regex` provides options for selecting different matching behaviors, such as ASCII character classes or Unicode scalar semantics, which corresponds more closely with other regex engines.
+
+## Detailed design
+
+First, we'll discuss the options that let you control a regex's behavior, and then explore the character classes that define the your pattern.
+
+### Options
+
+Options can be enabled and disabled in two different ways: as part of [regex internal syntax][internals], or applied as methods when declaring a `Regex`. For example, both of these `Regex`es are declared with case insensitivity:
+
+```swift
+let regex1 = /(?i)banana/
+let regex2 = Regex {
+    "banana"
+}.ignoringCase()`
+```
+
+Note that the `ignoringCase()` is available on any type conforming to `RegexComponent`, which means that you can use the more readable option-setting interface in conjunction with regex literals:
+
+```swift
+let regex3 = /banana/.ignoringCase()
+```
+
+Calling an option-setting method like `ignoringCase(_:)` acts like wrapping the callee in an option-setting group `(?:...)`. That is, while it sets the behavior for the callee, it doesn’t override options that are applied to more specific regions. In this example, the middle `"na"` in `"banana"` matches case-sensitively, despite the outer call to `ignoringCase()`:
+
+```swift
+let regex4 = Regex {
+    "ba"
+    "na".ignoringCase(false)
+    "na"
+}
+.ignoringCase()
+	
+"banana".contains(regex4)     // true
+"BAnaNA".contains(regex4)     // true
+"BANANA".contains(regex4)     // false
+
+// Equivalent to:
+let regex5 = /(?i)ba(?-i:na)na/
+```
+
+All option APIs are provided on `RegexComponent`, so they can be called on a `Regex` instance, or on any component that you would use inside a `RegexBuilder` block when the `RegexBuilder` module is imported.
+
+The options that `Regex` supports are shown in the table below. Options that affect _matching behavior_ are supported through both regex syntax and APIs, while options that have _structural_ or _syntactic_ effects are only supported through regex syntax.
+
+| **Matching Behavior**        |                |                               |
+|------------------------------|----------------|-------------------------------|
+| Case insensitivity           | `(?i)`         | `ignoringCase()`              |
+| Single-line mode             | `(?s)`         | `dotMatchesNewlines()`        |
+| Multi-line mode              | `(?m)`         | `anchorsMatchNewlines()`      |
+| ASCII-only character classes | `(?DSWP)`      | `usingASCIIDigits()`, etc     |
+| Unicode word boundaries      | `(?w)`         | `usingSimpleWordBoundaries()` |
```

Comment:
If we make this a binary toggle, we'll have to invent another switch if we ever want to support Unicode level 3. Should this take an enum instead? `.identifyingWordBoundaries(with: [.unicodeLevel[123], .defaultRule])` or similar?

Reply:
Implemented this as `wordBoundaryKind(_:)` — see https://github.com/apple/swift-experimental-string-processing/pull/295